internal/models/users/access_control/access_control.go

```diff
@@ -0,0 +1,175 @@
+	err := assignRoleToUserByEmail(db, email, user_roles.QuizAdmin)
+	if err == nil {
+		isActive = true
+	} else if err == errUserAlreadyHasRole {
+		return nil, ErrEmailAlreadyAdmin
+	} else {
+		if err == errNoUserFound {
+			err = preAssignRoleToUserByEmail(db, email, user_roles.QuizAdmin)
+			if err != nil {
+				return nil, err
+			}
+			isActive = false
+		} else {
+			return nil, err
+		}
+	}
+
+	return &UserAdmin{
+		Email:    email,
+		IsActive: isActive,
+		Role:     user_roles.QuizAdmin,
+	}, nil
+}
+
+var errNoUserFound = errors.New("no user with given email found")
+var errUserAlreadyHasRole = errors.New("user already has the given role")
+
+// Assigns a role to a user with the given email.
+func assignRoleToUserByEmail(db *sql.DB, email string, role user_roles.Role) error {
+	user, err := users.GetUserByEmail(db, email)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return errNoUserFound
+		} else {
+			return err
+		}
+	}
+
+	if user.Role == role {
+		return errUserAlreadyHasRole
+	}
+
+	result, err := db.Exec(`
+	UPDATE users
+	SET role=$1
+	WHERE email=$2;`, role.String(), email)
+	if err != nil {
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected < 1 {
+		return errNoUserFound
+	}
+
+	return err
+
+}
+
+// Preassigns a role to a user with the given email.
+func preAssignRoleToUserByEmail(db *sql.DB, email string, role user_roles.Role) error {
+	_, err := db.Exec(`
+	INSERT INTO preassigned_roles(email, role)
+	VALUES ($1, $2);`, email, role.String())
+	if err != nil {
+		// if unique constraint violated, the role is already preassigned
+		if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == "23505" {
+			return ErrEmailAlreadyAdmin
+		}
+	}
+	return err
+}
+
+var errNoPreassignedRoleForEmail = errors.New("no preassigned role for the given email")
+
+// Removes a preassigned QuizAdmin role from the given email.
+func removePreassignedAdmin(db *sql.DB, email string) error {
+	result, err := db.Exec(`
+	DELETE FROM preassigned_roles
+	WHERE email=$1;`, email)
+	if err != nil {
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected < 1 {
+		return errNoPreassignedRoleForEmail
+	}
+	return nil
+
+}
+
+var ErrNoAdminWithGivenEmail = errors.New("no admin with given email found")
+
+// Removes the QuizAdmin from the given email. If the user exists the role is revoked. If the user does not exist, the preassigned role is removed.
+//
+// Returns ErrNoAdminWithGivenEmail if there is no user or preassigned role for the given email.
+func RemoveAdminByEmail(db *sql.DB, email string) error {
+	err := assignRoleToUserByEmail(db, email, user_roles.User)
+	if err == nil {
+		return nil
+	} else if err == errUserAlreadyHasRole {
+		return ErrNoAdminWithGivenEmail
+	} else if err == errNoUserFound {
+		err = removePreassignedAdmin(db, email)
+		if err != nil {
+			if err == errNoPreassignedRoleForEmail {
```

Comment:
Instead of nesting if-statements do `err != nil && err == errNoPreassignedRoleForEmail` unless you plan on expanding this later.

Reply:
f296ea0dd09bbad631900192c85938c10b0c1759